File: exercise3/python/anc_samp.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import multivariate_normal

## Define distribution
mu = (np.array([1, 1]), np.array([-4, 3]))
Sigma = (2*np.eye(2), 4*np.eye(2))
pi_k = [0.3, 0.7]
K = 2

## Plot distribution
X, Y = np.meshgrid(np.linspace(-12.0, 7.0, 100), np.linspace(-6.0, 9.0, 100))
XY = np.stack((X.flatten(), Y.flatten()), axis=1)
Z = np.zeros(len(XY))
for k in range(K):
  Z += pi_k[k] * multivariate_normal.pdf(XY, mean=mu[k], cov=Sigma[k])

plt.contour(X, Y, Z.reshape(X.shape))

## Sample from distribution
N = 500
samples = np.zeros((N, 2))
for n in range(N):
  samples[n] = [0, 0] # XXX: ACTUALLY DRAW A SAMPLE HERE!!!

plt.plot(samples[:, 0], samples[:, 1], '.');

plt.show()

File: exercise3/matlab/anc_samp.m
%% Define distribution
mu = {[1, 1], [-4, 3]};
Sigma = {2*eye(2), 4*eye(2)};
pi_k = [0.3, 0.7];
K = 2;

%% Plot distribution
[X, Y] = meshgrid(linspace(-12, 7, 100), linspace(-6, 9, 100));
XY = [X(:), Y(:)];
Z = zeros(size(X));
for k = 1:K
  Z(:) = Z(:) + pi_k(k) * mvnpdf(XY, mu{k}, Sigma{k});
end % for
contour(X, Y, Z)

%% Sample from distribution
N = 500;
samples = NaN(N, 2);
for n = 1:N
  %samples(n, :) = NaN; % XXX: ACTUALLY DRAW A SAMPLE HERE!!!
  % First sample based on pi_k, then on the gaussians
  a = 0;
  b = 1;
  r = (b-a).*rand + a;
  if r<0.3
      index = 1;
  else
      index = 2;
  end
  samples(n, :) = (sqrt(Sigma{index})*randn(2,1)+mu{index}')';
  % WRONG
end % for
hold on
plot(samples(:, 1), samples(:, 2), '.');
hold off

